src/handformer.cpp

```diff
@@ -63,12 +72,89 @@ auto Mahjong::HandFromNotation(std::string notation) -> std::vector<Piece> {
         if (isError || i == -1) {
           return Piece(Piece::ERROR);
         }
+
         if (isHonor) {
-          return Piece(NOTATION_TO_HONOR[i]);
+          return Piece(NOTATION_TO_HONOR.at(i));
         }
-        return Piece::formPiece(setSuit, i);
+
+        auto isRedFive = i == 0;
+        auto value = isRedFive ? 5 : i;
+
+        return Piece::formPiece(setSuit, value, isRedFive);
       });
     currentTiles.clear();
   }
+  // Free Remaining tiles (in case of no ending suit)
+  currentTiles.clear();
+
   return result;
 }
+
+auto Mahjong::IsValidNotation(std::string notation) -> bool {
+  std::vector<uint8_t> currentSet;
+  for (const auto& c : notation) {
+    // If symbol, not valid
+    if (!std::isdigit(c) && !std::isalpha(c)) {
+      return false;
+    }
+
+    if (std::isdigit(c)) {
+      // Converts character digit to integer number (safe with UTF-7 and UTF-8)
+      auto value = c - '0';
+      currentSet.push_back(value);
+      continue;
+    }
+
+    // Suit Found, check if valid
+    if (NOTATION_TO_SUIT.count(c) == 0) {
+      return false;
+    }
+
+    if (c == 'z') {
+      for (const auto& i : currentSet) {
+        if (i > 7) {
+          return false;
+        }
+      }
+    }
+
+    currentSet.clear();
+  }
+
+  // Check if notation ended with suit for final set.
+  if (currentSet.size() > 0) {
+    return false;
+  }
+
+  return true;
+}
+
+auto Mahjong::HandToNotation(std::vector<Piece> hand) -> std::string {
+  std::map<Piece::Type, std::vector<uint8_t>> suitAndTiles;
+
+  for (const auto& t : hand) {
+    if (t == Piece(Piece::ERROR)) {
+      continue;
+    }
+
+    const auto suit = (Piece::Type)t.getSuit();
+    if (t.isRedFive()) {
+      suitAndTiles[suit].push_back(0);
+      continue;
+    }
+
+    suitAndTiles[suit].push_back(t.getPieceNum());
+  }
+
+  std::string result;
+  for (const auto& pair : suitAndTiles) {
+    std::string numberList;
+    for (const auto& i : pair.second) {
+      numberList.append(std::to_string(i));
+    }
+    result.append(numberList);
+    result.push_back(SUIT_TO_NOTATION.at(pair.first));
+  }
+
+  return result;
+}
```

Comment:
nit: /n

Reply:
also btw if you don't hate newlines at the end of all of your files there is vscode setting to auto append them on save

Reply:
fixed in 5ee5d6df3ae88121f47d7fc03d03b9139bcb821f